File: src/core/cpu.rs
use rand::Rng;

use crate::core::memory_bus::MemoryBus;

const REGISTER_SIZE: usize = 16;
const STACK_COUNT: usize = 12;
const PROGRAM_START_ADDRESS: u16 = 0x200;

pub struct CPU {
    registers: [u8; REGISTER_SIZE],
    i: u16,
    pc: u16,
    sp: usize,
    stack: [u16; STACK_COUNT],
    pub bus: MemoryBus
}

impl CPU {
    pub fn new() -> Self {
        Self {
            registers: [0; REGISTER_SIZE],
            i: 0,
            sp: 0,
            stack: [0; STACK_COUNT],
            bus: MemoryBus::new(),
            pc: PROGRAM_START_ADDRESS
        }
    }

    pub fn load_game(&mut self, file: &[u8]) {
        if Some(file).is_some() {
            if file.len() < 4096 - 512 {
                self.bus.memory[512..512+file.len()].copy_from_slice(&file[..]);
            }
        }
    }

    pub fn tick(&mut self) {
        let opcode = self.fetch();
        self.execute(opcode);
    }

    pub fn tick_timers(&mut self) {
        if self.bus.delay_timer > 0 {
            self.bus.delay_timer -= 1;
        }

        if self.bus.sound_timer > 0 {
            if self.bus.sound_timer == 1 {
                // BEEP
            }
            self.bus.sound_timer -= 1;
        }
    }

    fn fetch(&mut self) -> u16 {
        let higher_byte = self.bus.memory[self.pc as usize] as u16;
        let lower_byte = self.bus.memory[(self.pc + 1) as usize] as u16;
        let opcode = (higher_byte << 8) | lower_byte;
        self.pc += 2;
        opcode
    }


    fn execute(&mut self, opcode: u16) {
        let op1 = ((opcode & 0xF000) >> 12u8) as usize;
        let op2 = ((opcode & 0x0F00) >> 8u8) as usize;
        let op3 = ((opcode & 0x00F0) >> 4u8) as usize;
        let op4 = (opcode & 0x000F) as usize;

        match (op1, op2, op3, op4) {

            (0x0, 0x0, 0x0, 0x0) => return,
            (0x0, 0x0, 0xE, 0x0) => self.bus.graphics.memory = [false; 32 * 64],
            (0x0, 0x0, 0xE, 0xE) => {
                self.sp -= 2;
                self.pc = self.stack[self.sp];
            }
            (0x1, _, _, _) => {
                self.pc = opcode & 0x0FFF;
            }
            (0x2, _, _, _) => {
                self.stack[self.sp] = self.pc;
                self.sp += 2;
                self.pc = opcode & 0x0FFF;
            }
            (0x3, _, _, _) => {
                if self.registers[op2] == (opcode & 0x0FF) as u8 {
                    self.pc += 2;
                }
            }
            (0x4, _, _, _) => {
                if self.registers[op2] != (opcode & 0x0FF) as u8 {
                    self.pc += 2;
                }
            }
            (0x5, _, _, _) => {
                if self.registers[op2] == self.registers[op3]  {
                    self.pc += 2;
                }
            }
            (0x6, _, _, _) => {
                self.registers[op2] = (opcode & 0x00FF) as u8;
            }
            (0x7, _, _, _) => {
                let value = self.registers[op2].wrapping_add((opcode & 0x00FF) as u8);
                self.registers[op2] = value;
            }
            (0x8, _, _, 0x0) => {
                self.registers[op2] = self.registers[op3];
            }
            (0x8, _, _, 0x1) => {
                self.registers[op2] = self.registers[op2] | self.registers[op3];
            }
            (0x8, _, _, 0x2) => {
                self.registers[op2] &= self.registers[op3];
            }
            (0x8, _, _, 0x3) => {
                self.registers[op2] ^= self.registers[op3];
            }
            (0x8, _, _, 0x4) => {
                let (value, carry) = self.registers[op2].overflowing_add(self.registers[op3]);
                self.registers[op2] = value;
                self.registers[0xF] = carry as u8;;
            }
            (0x8, _, _, 0x5) => {
                let (value, borrow) = self.registers[op2].overflowing_sub(self.registers[op3]);
                self.registers[op2] = value;
                self.registers[0xF] = !borrow as u8;
            }
            (0x8, _, _, 0x6) => {
                let bit = self.registers[op3] & 0x1;
                self.registers[op2] = self.registers[op3] >> 1;
                self.registers[0xF] = bit;
            }
            (0x8, _, _, 0x7) => {
                let (value, borrow) = self.registers[op3].overflowing_sub(self.registers[op2]);
                self.registers[op2] = value;
                self.registers[0xF] = !borrow as u8;
            }
            (0x8, _, _, 0xE) => {
                let bit = (self.registers[op3] & 0x8) >> 3;
                self.registers[op2] = self.registers[op3] << 1;
                self.registers[0xF] = bit;
            }
            (0x9, _, _, 0x0) => {
                if self.registers[op2] != self.registers[op3] {
                    self.pc += 2;
                }
            }
            (0xA, _, _, _) => {
                self.i = opcode & 0x0FFF;
            }
            (0xB, _, _, _) => {
                self.pc = (opcode & 0x0FFF) + self.registers[0x0] as u16;
            }
            (0xC, _, _, _) => {
                let random = rand::thread_rng().gen_range(0..255);
                self.registers[op2] = random & (opcode & 0x00FF) as u8;
            }
            (0xD, _, _, _) => {
                let x = self.registers[op2] as u16;
                let y = self.registers[op3] as u16;
                let height = op4;

                self.registers[0xF] = 0;

                for y_line in 0..height as u16 {
                    let pixel = self.bus.memory[(self.i + y_line) as usize];

                    for x_line in 0..8 {
                        if pixel & (0x80 >> x_line) != 0 {
                            if self.bus.graphics.memory[(x + x_line + ((y + y_line) * 64)) as usize] {
                                self.registers[0xF] = 1;
                            }
                            self.bus.graphics.memory[(x + x_line + ((y + y_line) * 64)) as usize] = !self.bus.graphics.memory[(x + x_line + ((y + y_line) * 64)) as usize];
                        }
                    }

                    self.bus.graphics.draw = true;
                }
            }
            (0xE, _, 0x9, 0xE) => {
                if self.bus.input.keys[self.registers[op2] as usize] {
                    self.pc += 2;
                }
            }
            (0xE, _, 0xA, 0x1) => {
                if !self.bus.input.keys[self.registers[op2] as usize] {
                    self.pc += 2;
                }
            }
            (0xF, _, 0x0, 0x7) => {
                self.registers[op2] = self.bus.delay_timer;
            }
            (0xF, _, 0x0, 0xA) => {
                let mut key_pressed = false;

                for key in self.bus.input.keys {
                    if key {
                        key_pressed = true;
                    }

                    if !key_pressed {
                        self.pc -= 2;
                    }
                }
            }
            (0xF, _, 0x1, 0x5) => {
                self.bus.delay_timer = self.registers[op2];
            }
            (0xF, _, 0x1, 0x8) => {
                self.bus.sound_timer = self.registers[op2];
            }
            (0xF, _, 0x1, 0xE) => {
                self.i += self.registers[op2] as u16;
            }
            (0xF, _, 0x2, 0x9) => {
                self.i = self.registers[op2] as u16 * 0x5;
            }
            (0xF, _, 0x3, 0x3) => {
                self.bus.memory[self.i as usize] = (self.registers[op2] / 100);
                self.bus.memory[(self.i + 1) as usize] = ((self.registers[op2] / 10) % 10);
                self.bus.memory[(self.i + 2) as usize] = ((self.registers[op2] % 100) % 10);
            }
            (0xF, _, 0x5, 0x5) => {
                for i in 0..= op2 as u16 {
                    self.bus.memory[(self.i + i) as usize] = self.registers[i as usize];
                }
            }
            (0xF, _, 0x6, 0x5) => {
                for i in 0..= op2 as u16 {
                    self.registers[i as usize] = self.bus.memory[(self.i + i) as usize];
                }
            }
            (_, _, _, _) => unimplemented!("Unimplemented opcode: {}", opcode),
        }
    }
}

File: src/Sources/RetroC8/CPU.swift
//
//  File.swift
//  
//
//  Created by Glenn Hevey on 19/11/2023.
//

import Foundation

struct CPU {
    let registerSize: Int = 16
    let stackCount: Int = 12
    let programStartAddress: UInt16 = 0x200
    
    var i: UInt16
    var pc: UInt16
    var sp: UInt
    var registers: [UInt8]
    var stack: [UInt16]
    var bus: MemoryBus
    
    init() {
        i = 0
        sp = 0
        pc = programStartAddress
        registers = [UInt8](repeating: 0, count: registerSize)
        stack = [UInt16](repeating: 0, count: stackCount)
        bus = MemoryBus()
    }
    
    mutating func load(_ file: [UInt8]) {
        if file.count < 4096 - 512 {
            bus.memory.replaceSubrange(512...(512 + file.count), with: file)
        }
    }
    
    mutating func tick() {
        let opCode = fetch()
        execute(opCode: opCode)
    }
    
    mutating func tickTimers() {
        if bus.delayTimer > 0 {
            bus.delayTimer -= 1
        }
        
        if bus.soundTimer > 0 {
            if bus.soundTimer == 1 {
                //TODO: BEEP
            }
            bus.soundTimer -= 1
        }
    }
    
    mutating func fetch() -> UInt16 {
        let higherByte = bus.memory[Int(self.pc)]
        let lowerByte = bus.memory[Int(self.pc + 1)]
        
        let opCode: UInt16 = UInt16(higherByte) << 8 | UInt16(lowerByte)
        
        pc += 2
        
        return opCode
    }
    
    mutating func execute(opCode: UInt16) {
        let op1 = (opCode & 0xF000) >> 12
        let op2 = (opCode & 0x0F00) >> 8
        let op3 = (opCode & 0x00F0) >> 4
        let op4 = (opCode & 0x000F)
        
        switch (op1, op2, op3, op4) {
        case (0x0, 0x0, 0x0, 0x0):
            return
        case (0x0, 0x0, 0xE, 0x0):
            bus.gpu.memory = [Bool](repeating: false, count: bus.gpu.memory.count)
        case (0x0, 0x0, 0xE, 0xE):
            if sp >= 2 {
                sp -= 2
            }
            pc = stack[Int(sp)]
        case (0x1, _, _, _):
            pc = opCode & 0x0FFF
        case (0x2, _, _, _):
            stack[Int(sp)] = pc
            sp += 2
            pc = opCode & 0x0FFF
        case (0x3, _, _, _):
            if registers[Int(op2)] == opCode & 0x0FF {
                pc += 2
            }
        case (0x4, _, _, _):
            if registers[Int(op2)] != opCode & 0x0FF {
                pc += 2
            }
        case (0x5, _, _, _):
            if registers[Int(op2)] == registers[Int(op3)] {
                pc += 2
            }
        case (0x6, _, _, _):
            registers[Int(op2)] = UInt8(opCode & 0x00FF)
        case (0x7, _, _, _):
            let value = registers[Int(op2)] &+ UInt8(opCode & 0x00FF)
            registers[Int(op2)] = value
        case (0x8, _, _, 0x0):
            registers[Int(op2)] = registers[Int(op3)]
        case (0x8, _, _, 0x1):
            registers[Int(op2)] = registers[Int(op2)] | registers[Int(op3)]
            registers[0xF] = 0x0;
        case (0x8, _, _, 0x2):
            registers[Int(op2)] &= registers[Int(op3)]
            registers[0xF] = 0x0
        case (0x8, _, _, 0x3):
            registers[Int(op2)] ^= self.registers[Int(op3)]
            registers[0xF] = 0x0
        case (0x8, _, _, 0x4):
            let (value, carry) = registers[Int(op2)].addingReportingOverflow(registers[Int(op3)])
            registers[Int(op2)] = value
            registers[0xF] = carry == false ? 0x0 : 0xF
        case (0x8, _, _, 0x5):
            let (value, carry) = registers[Int(op2)].subtractingReportingOverflow(registers[Int(op3)])
            registers[Int(op2)] = value
            registers[0xF] = carry == true ? 0x0 : 0xF
        case (0x8, _, _, 0x6):
            let bit = registers[Int(op3)] & 0x1
            registers[Int(op2)] = registers[Int(op3)] >> 1
            registers[0xF] = bit
        case (0x8, _, _, 0x7):
            let (value, carry) = registers[Int(op3)].subtractingReportingOverflow(registers[Int(op2)])
            registers[Int(op2)] = value
            registers[0xF] = carry == true ? 0x0 : 0xF
        case (0x8, _, _, 0xE):
            let bit = (registers[Int(op3)] & 0x8) >> 3
            registers[Int(op2)] = registers[Int(op3)] << 1
            registers[0xF] = bit
        case (0x9, _, _, 0x0):
            if registers[Int(op2)] != registers[Int(op3)] {
                pc += 2
            }
        case (0xA, _, _, _):
            i = opCode & 0x0FFF
        case (0xB, _, _, _):
            pc = (opCode & 0x0FFF) + UInt16(registers[0x0])
        case (0xC, _, _, _):
            let random = UInt16.random(in: 0...255)
            registers[Int(op2)] = UInt8(random & (opCode & 0x00FF))
        case (0xD, _, _, _):
            let x = UInt16(registers[Int(op2)] % 64)
            let y = UInt16(registers[Int(op3)] % 32)
            let height = UInt16(op4)
            
            registers[0xF] = 0

            for y_line in 0..<height {
                let pixel = bus.memory[Int(i + y_line)]
                
                for x_line in 0..<UInt16(8) {
                    if x + x_line < 64 && y + y_line < 32 {
                        if pixel & (0x80 >> x_line) != 0 {
                            let location = x + x_line + ((y + y_line) * 64)  //UInt8(UInt(x + x_line) + (UInt((y + y_line)) * 64))
                            if bus.gpu.memory[Int(location)] {
                                registers[0xF] = 1
                            }
                            
                            let pixelValue = !bus.gpu.memory[Int(location)]
                            
                            bus.gpu.memory[Int(location)] = pixelValue
                        }
                    }
                }
                bus.gpu.draw = true
            }
        case (0xE, _, 0x9, 0xE):
            if bus.input.keys[Int(registers[Int(op2)])] {
                pc += 2
            }
        case (0xE, _, 0xA, 0x1):
            if !bus.input.keys[Int(registers[Int(op2)])] {
                pc += 2
            }
        case (0xF, _, 0x0, 0x7):
            registers[Int(op2)] = bus.delayTimer
        case (0xF, _, 0x0, 0xA):
            var keyPressed = false
            
            for key in bus.input.keys {
                if key {
                    keyPressed = true
                }
                
                if !keyPressed {
                    pc -= 2
                }
            }
        case (0xF, _, 0x1, 0x5):
            bus.delayTimer = registers[Int(op2)]
        case (0xF, _, 0x1, 0x8):
            bus.soundTimer = registers[Int(op2)]
        case (0xF, _, 0x1, 0xE):
            i += UInt16(registers[Int(op2)])
        case (0xF, _, 0x2, 0x9):
            i = UInt16(registers[Int(op2)]) * 0x5
        case (0xF, _, 0x3, 0x3):
            bus.memory[Int(i)] = registers[Int(op2)] / 100
            bus.memory[Int(i + 1)] = (registers[Int(op2)] / 10) % 10
            bus.memory[Int(i + 2)] = (registers[Int(op2)] % 100) % 10
        case (0xF, _, 0x5, 0x5):
            var i_temp: UInt16 = 0
            for j in 0...op2 {
                bus.memory[Int(i + j)] = registers[Int(j)]
                i_temp = j
            }
            self.i = i_temp + UInt16(op2) + 1
        case (0xF, _, 0x6, 0x5):
            var i_temp: UInt16 = 0
            for j in 0...op2 {
                registers[Int(j)] = bus.memory[Int(i + j)]
                i_temp = j
            }
            self.i = i_temp + UInt16(op2) + 1;
        default:
            fatalError("Unimplemented OpCode: \(opCode)")
        }
    }
    

}


